src/mpn_mod.h

```diff
@@ -0,0 +1,339 @@
+{
+    if (flint_mpn_zero_p(x, n))
+        flint_mpn_zero(res, n);
+    else
+        mpn_sub_n(res, m, x, n);
+}
+
+FLINT_FORCE_INLINE void
+flint_mpn_addmod_n(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_srcptr m, mp_size_t n)
+{
+    mp_limb_t cy;
+    cy = mpn_add_n(res, x, y, n);
+    if (cy || mpn_cmp(res, m, n) >= 0)
+        mpn_sub_n(res, res, m, n);
+}
+
+FLINT_FORCE_INLINE void
+flint_mpn_submod_n(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_srcptr m, mp_size_t n)
+{
+    int cmp = (mpn_cmp(x, y, n) < 0);
+    mpn_sub_n(res, x, y, n);
+    if (cmp)
+        mpn_add_n(res, res, m, n);
+}
+
+/* assumes m <= n and y < m */
+FLINT_FORCE_INLINE void
+flint_mpn_addmod_n_m(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_size_t yn, mp_srcptr m, mp_size_t n)
+{
+    mp_limb_t cy;
+    cy = mpn_add(res, x, n, y, yn);
+    if (cy || mpn_cmp(res, m, n) >= 0)
+        mpn_sub_n(res, res, m, n);
+}
+
+/* assumes m <= n and y < m */
+FLINT_FORCE_INLINE void
+flint_mpn_submod_n_m(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_size_t yn, mp_srcptr m, mp_size_t n)
+{
+    int cmp = (flint_mpn_zero_p(x + yn, n - yn) && mpn_cmp(x, y, yn) < 0);
+    mpn_sub(res, x, n, y, yn);
+    if (cmp)
+        mpn_add_n(res, res, m, n);
+}
+
+FLINT_FORCE_INLINE void
+flint_mpn_negmod_2(mp_ptr res, mp_srcptr x, mp_srcptr m)
+{
+    if (x[0] == 0 && x[1] == 0)
+        res[1] = res[0] = 0;
+    else
+        sub_ddmmss(res[1], res[0], m[1], m[0], x[1], x[0]);
+}
+
+FLINT_FORCE_INLINE void
+flint_mpn_addmod_2(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_srcptr m)
+{
+    mp_limb_t cy;
+    mp_limb_t m1 = m[1], m0 = m[0];
+    add_sssaaaaaa(cy, res[1], res[0], 0, x[1], x[0], 0, y[1], y[0]);
+    if (cy || (res[1] > m1 || (res[1] == m1 && res[0] >= m0)))
+        sub_ddmmss(res[1], res[0], res[1], res[0], m1, m0);
+}
+
+FLINT_FORCE_INLINE void
+flint_mpn_submod_2(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_srcptr m)
+{
+    int cmp;
+    mp_limb_t m1 = m[1], m0 = m[0];
+    mp_limb_t x1 = x[1], x0 = x[0];
+    mp_limb_t y1 = y[1], y0 = y[0];
+    cmp = (x1 < y1) || (x1 == y1 && x0 < y0);
+    sub_ddmmss(res[1], res[0], x1, x0, y1, y0);
+    if (cmp)
+        add_ssaaaa(res[1], res[0], res[1], res[0], m1, m0);
+}
+
+FLINT_FORCE_INLINE int
+flint_mpn_signed_sub_n(mp_ptr res, mp_srcptr x, mp_srcptr y, mp_size_t n)
+{
+    if (mpn_cmp(x, y, n) >= 0)
+    {
+        mpn_sub_n(res, x, y, n);
+        return 0;
+    }
+    else
+    {
+        mpn_sub_n(res, y, x, n);
+        return 1;
+    }
+}
+
+FLINT_FORCE_INLINE void
+flint_mpn_signed_div2(mp_ptr res, mp_srcptr x, mp_size_t n)
+{
+    mp_limb_t s = x[n - 1] & (UWORD(1) << (FLINT_BITS - 1));
+    mpn_rshift(res, x, n, 1);
+    res[n - 1] |= s;
+}
+
+void flint_mpn_mulmod_preinvn_2(mp_ptr r,
+        mp_srcptr a, mp_srcptr b,
+        mp_srcptr d, mp_srcptr dinv, ulong norm);
+
+char * _flint_mpn_get_str(mp_srcptr x, mp_size_t n);
+
+/* Basic operations and arithmetic */
+
+int gr_ctx_init_mpn_mod(gr_ctx_t ctx, const fmpz_t n);
```

Comment:
From the low-level perspective, it doesn't really make sense to have an fmpz as input here. Can we at least add an underscored version that takes an `mp_srcptr` along with `mp_size_t`?

Reply:
Yes, I was planning to do that.